receiver/aerospikereceiver/scraper.go

```diff
@@ -0,0 +1,217 @@
+// Copyright 2022, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aerospikereceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/aerospikereceiver"
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strconv"
+	"time"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/aerospikereceiver/internal/metadata"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/aerospikereceiver/internal/model"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/receiver/scrapererror"
+	"go.uber.org/zap"
+)
+
+// aerospikeReceiver is a metrics receiver using the Aerospike interface to collect
+type aerospikeReceiver struct {
+	params   component.ReceiverCreateSettings
+	config   *Config
+	consumer consumer.Metrics
+	host     string // host/IP of configured Aerospike node
+	port     int    // port of configured Aerospike node
+	mb       *metadata.MetricsBuilder
+	logger   *zap.Logger
+}
+
+// newAerospikeReceiver creates a new aerospikeReceiver connected to the endpoint provided in cfg
+//
+// If the host or port can't be parsed from endpoint, an error is returned.
+func newAerospikeReceiver(params component.ReceiverCreateSettings, cfg *Config, consumer consumer.Metrics) (*aerospikeReceiver, error) {
+	host, portStr, err := net.SplitHostPort(cfg.Endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("endpoint: %w", err)
+	}
+
+	port, err := strconv.ParseInt(portStr, 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("port: %w", err)
+	}
+
+	return &aerospikeReceiver{
+		params:   params,
+		config:   cfg,
+		consumer: consumer,
+		host:     host,
+		port:     int(port),
+		mb:       metadata.NewMetricsBuilder(cfg.Metrics),
+		logger:   params.Logger.Named("aerospikereceiver"),
+	}, nil
+}
+
+// scrape scrapes both Node and Namespace metrics from the provided Aerospike node.
+// If CollectClusterMetrics is true, it then scrapes every discovered node
+func (r *aerospikeReceiver) scrape(ctx context.Context) (pmetric.Metrics, error) {
+	errs := &scrapererror.ScrapeErrors{}
+	now := pcommon.NewTimestampFromTime(time.Now().UTC())
+	client, err := newASClient(r.host, r.port, r.config.Timeout)
```

Comment:
Do we need to create a new client every time? Or is it possible to initialize it once in a Start function, and keep it for the lifetime of the receiver?

Reply:
It is, but not if `collect_cluster_metrics` is true since we dynamically connect to each node

Reply:
Makes sense; I asked because it makes it easier to unit test if you can just attach a single mock client.